components/instance-creator/internal/client/client.go

```diff
@@ -0,0 +1,446 @@
+		if err != nil {
+			return errors.Wrapf(err, "while building %s URL", resources.GetType())
+		}
+
+		req, err := http.NewRequest(http.MethodDelete, strURL, nil)
+		if err != nil {
+			return err
+		}
+
+		log.C(ctx).Infof("Deleting record of %s with ID: %q for subaccount: %q", resources.GetType(), resourceID, subaccountID)
+		resp, err := caller.Call(req)
+		if err != nil {
+			return err
+		}
+		defer closeResponseBody(ctx, resp)
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return errors.Errorf("failed to read response body from %s deletion request: %v", resources.GetType(), err)
+		}
+
+		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusAccepted {
+			return errors.Errorf("failed to delete a record of %s, status: %d, body: %s", resources.GetType(), resp.StatusCode, body)
+		}
+
+		if resp.StatusCode == http.StatusAccepted {
+			log.C(ctx).Infof("Handle asynchronous %s deletion...", resources.GetType())
+			_, err := c.executeAsyncRequest(ctx, resp, caller, subaccountID, false)
+			if err != nil {
+				return errors.Wrapf(err, "while deleting a record of %s with ID: %q for subaccount: %q", resources.GetType(), resourceID, subaccountID)
+			}
+			return nil
+		}
+
+		log.C(ctx).Infof("Successfully deleted a record of %s with ID: %q for subaccount: %q synchronously", resources.GetType(), resourceID, subaccountID)
+	}
+
+	return nil
+}
+
+func buildURL(baseURL, path, tenantKey, tenantValue string) (string, error) {
+	base, err := url.Parse(baseURL)
+	if err != nil {
+		return "", err
+	}
+
+	// Path params
+	base.Path += path
+
+	// Query params
+	params := url.Values{}
+	params.Add(tenantKey, tenantValue)
+	base.RawQuery = params.Encode()
+
+	return base.String(), nil
+}
+
+func closeResponseBody(ctx context.Context, resp *http.Response) {
+	if err := resp.Body.Close(); err != nil {
+		log.C(ctx).Errorf("An error has occurred while closing response body: %v", err)
+	}
+}
+
+func (c *client) executeSyncRequest(ctx context.Context, strURL, region string) ([]byte, error) {
+	req, err := http.NewRequest(http.MethodGet, strURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	caller, err := c.callerProvider.GetCaller(c.cfg, region)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting caller for region: %s", region)
+	}
+
+	resp, err := caller.Call(req)
+	if err != nil {
+		return nil, err
+	}
+	defer closeResponseBody(ctx, resp)
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Errorf("failed to read object, response body: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("failed to get object(s), status: %d, body: %s", resp.StatusCode, body)
+	}
+
+	return body, nil
+}
+
+func (c *client) executeAsyncRequest(ctx context.Context, resp *http.Response, caller ExternalSvcCaller, subaccountID string, isCreateRequest bool) (string, error) {
```

Comment:
Why did we choose the polling approach, was it not possible to be notified via callback? If we wait for the async resource processing we risk reaching the timeout of the client that is calling us or the server timeout. Can we push the async operations in some queue and return something that notifies that. Then when all of the operations are processed we could aggregate the resources and then notify the caller asynchronously with the information